File: apps/syn-cli/src/syn_cli/commands/agent.py
"""Agent interaction commands — list providers, test, chat."""

from __future__ import annotations

from typing import Annotated

import typer
from rich.table import Table

from syn_cli._output import console, print_error
from syn_cli.client import get_client

app = typer.Typer(
    name="agent",
    help="AI agent management and testing",
    no_args_is_help=True,
)


def _handle_connect_error() -> None:
    from syn_cli.client import get_api_url

    print_error(f"Could not connect to API at {get_api_url()}")
    console.print("[dim]Make sure the API server is running.[/dim]")
    raise typer.Exit(1)


@app.command("list")
def list_providers() -> None:
    """List available agent providers."""
    try:
        with get_client() as client:
            resp = client.get("/agents/providers")
    except Exception:
        _handle_connect_error()
        return

    if resp.status_code != 200:
        print_error(resp.json().get("detail", f"HTTP {resp.status_code}"))
        raise typer.Exit(1)

    providers = resp.json()
    if not isinstance(providers, list):
        providers = providers.get("providers", [])

    table = Table(title="Agent Providers")
    table.add_column("Provider", style="cyan")
    table.add_column("Display Name")
    table.add_column("Available")
    table.add_column("Default Model")

    for p in providers:
        available = "[green]Yes[/green]" if p.get("available") else "[red]No[/red]"
        table.add_row(
            p.get("provider", ""),
            p.get("display_name", ""),
            available,
            p.get("default_model", ""),
        )
    console.print(table)


@app.command("test")
def test_agent(
    provider: Annotated[
        str,
        typer.Option("--provider", "-p", help="Agent provider (claude, mock)"),
    ] = "claude",
    prompt: Annotated[
        str,
        typer.Option("--prompt", help="Test prompt"),
    ] = "Say hello in one sentence.",
    model: Annotated[
        str | None,
        typer.Option("--model", "-m", help="Model override"),
    ] = None,
) -> None:
    """Test an agent provider with a simple prompt."""
    try:
        with get_client() as client, console.status(f"Testing {provider}..."):
            resp = client.post(
                "/agents/test",
                json={"provider": provider, "prompt": prompt, "model": model},
                timeout=60.0,
            )
    except Exception:
        _handle_connect_error()
        return

    if resp.status_code != 200:
        print_error(resp.json().get("detail", f"HTTP {resp.status_code}"))
        raise typer.Exit(1)

    result = resp.json()
    console.print(f"[green]Response from {result.get('provider', provider)}:[/green]")
    console.print(f"  Model: {result.get('model', 'unknown')}")
    console.print(f"  Response: {result.get('response_text', '')}")
    console.print(
        f"  Tokens: {result.get('input_tokens', 0)} in / {result.get('output_tokens', 0)} out"
    )


@app.command("chat")
def chat_session(
    provider: Annotated[
        str,
        typer.Option("--provider", "-p", help="Agent provider"),
    ] = "claude",
    system: Annotated[
        str | None,
        typer.Option("--system", "-s", help="System prompt"),
    ] = None,
    model: Annotated[
        str | None,
        typer.Option("--model", "-m", help="Model override"),
    ] = None,
) -> None:
    """Start an interactive chat session."""
    console.print(f"[bold]Chat with {provider}[/bold] (type 'exit' to quit)\n")

    messages: list[dict[str, str]] = []
    if system:
        messages.append({"role": "system", "content": system})

    while True:
        try:
            user_input = console.input("[bold blue]You:[/bold blue] ")
        except (EOFError, KeyboardInterrupt):
            console.print("\n[dim]Goodbye![/dim]")
            break

        if user_input.strip().lower() in ("exit", "quit", "q"):
            console.print("[dim]Goodbye![/dim]")
            break

        if not user_input.strip():
            continue

        messages.append({"role": "user", "content": user_input})

        try:
            with get_client() as client, console.status("Thinking..."):
                resp = client.post(
                    "/agents/chat",
                    json={"provider": provider, "messages": messages, "model": model},
                    timeout=120.0,
                )
        except Exception:
            print_error("Lost connection to API server")
            break

        if resp.status_code != 200:
            print_error(resp.json().get("detail", f"HTTP {resp.status_code}"))
            break

        result = resp.json()
        response_text = result.get("response_text", "")
        console.print(f"[bold green]Agent:[/bold green] {response_text}\n")
        messages.append({"role": "assistant", "content": response_text})

File: apps/syn-cli-node/src/commands/agent.ts
/**
 * Agent interaction commands — list providers, test.
 * Port of apps/syn-cli/src/syn_cli/commands/agent.py
 * Note: chat command omitted (requires interactive stdin, not suited for non-interactive CLI).
 */

import { CommandGroup, type CommandDef, type ParsedArgs } from "../framework/command.js";
import { apiGetList, apiPost } from "../client/api.js";
import { print } from "../output/console.js";
import { style, CYAN, GREEN } from "../output/ansi.js";
import { Table } from "../output/table.js";

const listCommand: CommandDef = {
  name: "list",
  description: "List available agent providers",
  handler: async () => {
    const providers = await apiGetList<Record<string, unknown>>("/agents/providers");

    const table = new Table({ title: "Agent Providers" });
    table.addColumn("Provider", { style: CYAN });
    table.addColumn("Display Name");
    table.addColumn("Available");
    table.addColumn("Default Model");

    for (const p of providers) {
      const available = p["available"]
        ? style("Yes", GREEN)
        : style("No", "\x1b[31m");
      table.addRow(
        String(p["provider"] ?? ""),
        String(p["display_name"] ?? ""),
        available,
        String(p["default_model"] ?? ""),
      );
    }
    table.print();
  },
};

const testCommand: CommandDef = {
  name: "test",
  description: "Test an agent provider with a simple prompt",
  options: {
    provider: { type: "string", short: "p", description: "Agent provider (claude, mock)", default: "claude" },
    prompt: { type: "string", description: "Test prompt", default: "Say hello in one sentence." },
    model: { type: "string", short: "m", description: "Model override" },
  },
  handler: async (parsed: ParsedArgs) => {
    const provider = (parsed.values["provider"] as string | undefined) ?? "claude";
    const prompt = (parsed.values["prompt"] as string | undefined) ?? "Say hello in one sentence.";
    const model = (parsed.values["model"] as string | undefined) ?? null;

    const result = await apiPost<Record<string, unknown>>("/agents/test", {
      body: { provider, prompt, model },
      timeoutMs: 60_000,
    });

    print(`${style(`Response from ${String(result["provider"] ?? provider)}:`, GREEN)}`);
    print(`  Model: ${String(result["model"] ?? "unknown")}`);
    print(`  Response: ${String(result["response_text"] ?? "")}`);
    print(`  Tokens: ${String(result["input_tokens"] ?? 0)} in / ${String(result["output_tokens"] ?? 0)} out`);
  },
};

export const agentGroup = new CommandGroup("agent", "AI agent management and testing");
agentGroup.command(listCommand).command(testCommand);
